File: lib-segment/src/main/java/websiteschema/mpsegment/pinyin/WordToPinyinModel.java
package websiteschema.mpsegment.pinyin;

import websiteschema.mpsegment.hmm.*;
import websiteschema.mpsegment.util.SerializeHandler;

import java.io.*;
import java.util.Map;

public class WordToPinyinModel {

    private Viterbi viterbi = new Viterbi();
    private NodeRepository stateBank = new NodeRepository();
    private NodeRepository observeBank = new NodeRepository();
    private Trie ngram = new Trie();
    private Pi pi = new Pi();
    private Emission emission = new Emission();

    public boolean knwonObserve(String ch) {
        return observeBank.get(ch) != null;
    }

    public void load(String filename) throws IOException {
        File file = new File(filename);
        InputStream is = null;
        if (!file.exists()) {
            is = this.getClass().getClassLoader().getResourceAsStream("websiteschema/mpsegment/wtp.m");
            load(is);
        } else {
            is = new FileInputStream(file);
            load(is);
        }
    }

    public void load(InputStream is) {
        try {
            SerializeHandler handler = new SerializeHandler(new DataInputStream(is));
            observeBank.load(handler);
            stateBank.load(handler);
            pi.load(handler);
            emission.load(handler);
            ngram.load(handler);
            buildViterbi();
            handler.close();
        } catch (IOException ex) {
            ex.printStackTrace();
        }
    }

    public void save(String file) throws IOException {
        SerializeHandler handler = new SerializeHandler(new File(file), SerializeHandler.MODE_WRITE_ONLY);
        observeBank.save(handler);
        stateBank.save(handler);
        pi.save(handler);
        emission.save(handler);
        ngram.save(handler);
        handler.close();
    }


    protected Viterbi getViterbi() {
        return viterbi;
    }

    protected Trie getNgram() {
        return ngram;
    }

    protected NodeRepository getStateBank() {
        return stateBank;
    }

    protected NodeRepository getObserveBank() {
        return observeBank;
    }

    protected void buildPi(Map<Integer, Integer> pii) {
        pi = new Pi(pii);
    }

    protected void buildEmission(Map<Integer, Map<Integer, Integer>> emisMatrix) {
        emission = new Emission(emisMatrix);
    }

    protected void buildViterbi() {
        viterbi.setObserveBank(observeBank);
        viterbi.setStateBank(stateBank);
        viterbi.setPi(pi);
        Transition tran = new Transition(ngram, stateBank);
        viterbi.setTran(tran);
        viterbi.setE(emission);
    }
}

File: lib-segment-scala/src/main/scala/websiteschema/mpsegment/pinyin/WordToPinyinModel.scala
package websiteschema.mpsegment.pinyin

import websiteschema.mpsegment.hmm._
import websiteschema.mpsegment.util.SerializeHandler

import collection.mutable.Map
import java.io.{DataInputStream, InputStream, FileInputStream, File}
import websiteschema.mpsegment.util.FileUtil._
import io.Source

class WordToPinyinModel {

  private val viterbi = new Viterbi()
  private val stateBank = new NodeRepository()
  private val observeBank = new NodeRepository()
  private val ngram = new Trie()
  private var pi = Pi()
  private var emission: Emission = new Emission()

  def containsObserve(ch: String): Boolean = {
    return observeBank.get(ch) != null
  }

  def load(filename: String) {
    load(getResourceAsStream(filename))
  }

  def load(is: InputStream) {
    try {
      val handler = SerializeHandler(new DataInputStream(is))
      observeBank.load(handler)
      stateBank.load(handler)
      pi.load(handler)
      emission.load(handler)
      ngram.load(handler)
      buildViterbi()
      handler.close()
    } catch {
      case ex: Throwable =>
        ex.printStackTrace()
    }
  }

  def loadDictionary(dictFile : String) {
    val source = Source.fromInputStream(getResourceAsStream(dictFile))
    for (line <- source.getLines()) {
      val regex = "([^ ]+) *= *([^ ]+)".r
      regex findFirstIn line match {
        case Some(regex(han, pinyin)) => addPinyin(han, pinyin)
        case None =>
      }
    }
  }


  def addPinyin(han: String, pinyin: String) {
    val observe = observeBank.add(Node(han))
    val state = stateBank.get(pinyin)
    if (state != null) {
      emission.setProb(state.getIndex, observe.getIndex, 0.5D)
    }
  }

  def save(file: String) {
    val handler = SerializeHandler(new File(file), SerializeHandler.MODE_WRITE_ONLY)
    observeBank.save(handler)
    stateBank.save(handler)
    pi.save(handler)
    emission.save(handler)
    ngram.save(handler)
    handler.close()
  }


  def getViterbi(): Viterbi = {
    return viterbi
  }

  def getNgram(): Trie = {
    return ngram
  }

  def getStateBank(): NodeRepository = {
    return stateBank
  }

  def getObserveBank(): NodeRepository = {
    return observeBank
  }

  def buildPi(pii: Map[Int, Int]) {
    pi = Pi(pii)
  }

  def buildEmission(emisMatrix: Map[Int, Map[Int, Int]]) {
    emission = Emission(emisMatrix)
  }

  def buildViterbi() {
    viterbi.setObserveBank(observeBank)
    viterbi.setStateBank(stateBank)
    viterbi.setPi(pi)
    var tran = Transition(ngram, stateBank)
    viterbi.setTran(tran)
    viterbi.setE(emission)
  }
}